cloud_Azure/terraform/module/examples/multiple_accounts_multiple_resource_group/azure_cli.py

```diff
@@ -0,0 +1,46 @@
+import logging
+import os
+import time
+from typing import Any, Optional
+
+from az.cli import az
+
+log = logging.getLogger(__name__)
+
+
+def az_cli(command: str, max_attempts: int = 1) -> Optional[Any]:
+    """
+    Azure CLI commands issued in a quick succession may fail,
+    eg. when trying to configure a resource that is still being created
+    So, allow optional retry with linear backoff
+    """
+
+    for attempt_no in range(max_attempts):
+        data = _az_cli(command)
+        if data is not None:
```

Comment:
A simpler way to write the retry loop:

```python
wait_sec = 0
while max_attempts:
    data = _az_cli(command)
    if data is not None: 
        return data
    max_attempts -= 1
    wait_sec += <min_wait_time>
    log.debug("Retrying Azure CLI command in %d second(s)... (remaining attempts: %d)", wait_sec, max_attempts)
    time.sleep(wait_sec)
```

Reply:
az(...) calls return empty dict in case the was no data outputted.
So it is safe to log error and return None on error.

Reply:
We can consider migrating to azure sdk, but such process will take time, while working solution based on az cli is almost ready

Reply:
Relying on fixed number of waits and retries to let asynchronous operations to complete does not give me the right warm fuzzies. Can you, please, try one of the operations that require this treatment with Azure SDK to see if it is any better.

Was there a specific reason why you have chosen to use `az.cli` instead of the SDK or was it just random choice (first found, first used) :). I don't have any experience with either, but `az.cli` looks fragile to me, due to the need to construct command strings and then (at least in some cases) parse output. Also, the very large number of dependencies is unpleasant. So, I'd prefer to migrate to the SDK, but it does not need to be immediately. We can put it on the roadmap and do it when time permits.

Reply:
Sure, i can see what AzureSDK has to offer.
The reason is that az cli is anyway already used by terraform configuration itself (providers.tf, get_nsg.py), so why mixing another tool into the solution.